optimum/habana/transformers/models/mpt/modeling_mpt.py

```diff
@@ -15,180 +15,207 @@
-
-
-def gaudi_mpt_block_forward(
-    self,
-    hidden_states: torch.Tensor,
-    position_bias: torch.Tensor,
-    attention_mask: torch.Tensor,
-    layer_past: Optional[Tuple[torch.Tensor, torch.Tensor]] = None,
-    use_cache: bool = False,
-    output_attentions: bool = False,
-    token_idx: Optional[torch.Tensor] = None,
-    use_flash_attention: Optional[bool] = False,
-    flash_attention_recompute: Optional[bool] = False,
-):
-    """
-    Copied from MptBlock.forward: https://github.com/huggingface/transformers/blob/v4.32.0/src/transformers/models/mpt/modeling_mpt.py
-    The only differences are:
-    - add new args token_idx
-    - add new args use_flash_attention
-    - add new arg flash_attention_recompute
-    """
-    # hidden_states: [batch_size, seq_length, hidden_size]
-    # Layer norm at the beginning of the transformer layer.
-    layernorm_output = self.norm_1(hidden_states)
-
-    residual = hidden_states
-
-    # Self attention.
-    attn_outputs, attn_weights, past_key_value = self.attn(
-        layernorm_output,
-        position_bias=position_bias,
-        attention_mask=attention_mask,
-        past_key_value=layer_past,
-        token_idx=token_idx,
-        use_flash_attention=use_flash_attention,
-        flash_attention_recompute=flash_attention_recompute,
-    )
-
-    hidden_states = self.resid_attn_dropout(attn_outputs) + residual
-
-    layernorm_output = self.norm_2(hidden_states)
-
-    # Get residual
-    residual = hidden_states
-
-    # MLP.
-    output = self.ffn(layernorm_output, residual)
-    outputs = (output,)
-
-    if use_cache:
-        outputs += (past_key_value,)
-
-    if output_attentions:
-        outputs += (attn_weights,)
-
-    return outputs  # hidden_states, present, attentions
+            if len(position_bias.shape) != 3:
+                raise ValueError(f"Expecting position_bias shape to be 3 dimensions, got {len(position_bias.shape)}")
+            key_length = key_states.shape[-2]
+
+            position_bias_query_index = max(0, position_bias.size(1) - query_length)
+            position_bias_key_index = max(0, position_bias.size(2) - key_length)
+
+            position_bias = position_bias[:, position_bias_query_index:, position_bias_key_index:]
+
+        if use_flash_attention and FusedSDPA:
+            import habana_frameworks.torch.hpu as ht
+
+            with ht.sdp_kernel(enable_recompute=flash_attention_recompute):
+                attn_output = FusedSDPA.apply(
+                    query_states,
+                    key_states,
+                    value_states,
+                    attention_mask * torch.finfo(query_states.dtype).min + position_bias.to(query_states.dtype),
+                    0.0,
+                    False,
+                    None,
+                )
+            attn_weights = None
+        else:
+            attention_scores = torch.matmul(query_states, key_states.transpose(-1, -2)) * self.softmax_scale
+
```

Comment:
You should also use fp8 matmul kernel. Please check falcon code below. 
https://github.com/huggingface/optimum-habana/blob/main/optimum/habana/transformers/models/falcon/modeling_falcon.py#L108

Reply:
I tested it and throughput slowdown. The trace showed if MPT use it then it doesn't use fp8, but torch.matmul use fp8. That's the why MPT doesn't use the fp8 matmul kernel this time.

Reply:
That's weird.  Wondering if fp8_sofmax actually forcing matmul to also run with fp8.  Do you see a good accuracy with fp8 softmax?

Reply:
```
class Matmul(torch.nn.Module):
    def __init__(self):
        super().__init__()

    def forward(self, x, y):
        return torch.matmul(x, y)
```

Compare profile data this between FP8, for GPT-J, Mistral and MPT.

- GPT-J and Mistral spend <70% for MME but MPT spend ~30%.
- GPT-J and Mistral calling index_copy_fwd_hf8 and cast_bf16_to_hf8 kernels but doesn't appear these kernel call in MPT. When add matmul_qk and matmul_qv in blocklist of maxabs_quant.json for MPT then it call same kernels of GPT-J/Mistral.
This is reason we didn't add this.

Reply:
Also, can you make sure your change doesn't break anything for the training case since model file is used for both.